server/s3_put_multiobject_copy_action.cc

```diff
@@ -0,0 +1,962 @@
+      std::bind(&S3PutMultipartCopyAction::save_metadata_failed, this));
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::save_object_metadata_success() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  s3_copy_part_action_state = S3CopyPartActionState::metadataSaved;
+  next();
+}
+
+void S3PutMultipartCopyAction::save_metadata_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  s3_copy_part_action_state = S3CopyPartActionState::metadataSaveFailed;
+  if (part_metadata->get_state() == S3PartMetadataState::failed_to_launch) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Save of Part metadata failed due to pre launch failure\n");
+    set_s3_error("ServiceUnavailable");
+  } else {
+    s3_log(S3_LOG_ERROR, request_id, "Save of Part metadata failed\n");
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+const char xml_decl[] = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
+const char xml_comment_begin[] = "<!--   \n";
+const char xml_comment_end[] = "\n   -->\n";
+
+std::string S3PutMultipartCopyAction::get_response_xml() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  std::string xml_response = "";
+  xml_response += "<CopyPartResult>\n";
+  xml_response += S3CommonUtilities::format_xml_string(
+      "ETag", motr_writer->get_content_md5(), true);
+  xml_response += S3CommonUtilities::format_xml_string(
+      "LastModified", additional_object_metadata->get_last_modified_iso());
+  xml_response += "\n</CopyPartResult>";
+  s3_log(S3_LOG_INFO, stripped_request_id, "Base xml: %s", xml_response.c_str());
+
+  return xml_response;
+}
+
+void S3PutMultipartCopyAction::start_response() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  request->set_out_header_value("Content-Type", "application/xml");
+  request->set_out_header_value("Connection", "close");
+
+  request->send_reply_start(S3HttpSuccess200);
+
+  response_started = true;
+  request->set_response_started_by_action(true);
+}
+
+bool S3PutMultipartCopyAction::if_source_and_destination_same() {
+  return ((additional_bucket_name == request->get_bucket_name()) &&
+          (additional_object_name == request->get_object_name()));
+}
+
+void S3PutMultipartCopyAction::add_object_oid_to_probable_dead_oid_list() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  std::map<std::string, std::string> probable_oid_list;
+  assert(!new_oid_str.empty());
+
+  // store old object oid
+  if (old_object_oid.u_hi || old_object_oid.u_lo) {
+    assert(!old_oid_str.empty());
+
+    // prepending a char depending on the size of the object (size based
+    // bucketing of object)
+    S3CommonUtilities::size_based_bucketing_of_objects(
+        old_oid_str, part_metadata->get_content_length());
```

Comment:
double check here too, if get_content_length (i.e request content) should be set here?

Reply:
It should be for simple objects, for range based copy, we will need to update it with size specified by range.